File: packages/ios/Sources/CivUI/CivFieldset.swift
// CivUI — CivFieldset for SwiftUI
// Accessible fieldset container following government design system patterns.
// Renders: legend → hint → error → content (Section 508 compliant)

import SwiftUI

/// Accessible fieldset container for government applications.
///
/// GroupBox-like container with legend, hint, error, and disabled cascade.
/// Uses `@ViewBuilder` for arbitrary content.
///
/// Usage:
/// ```swift
/// CivFieldset(legend: "Mailing address") {
///     CivTextInput(label: "Street address", value: $street)
///     CivTextInput(label: "City", value: $city)
///     CivTextInput(label: "State", value: $state)
///     CivTextInput(label: "ZIP code", value: $zip, mask: .zip)
/// }
/// ```
public struct CivFieldset<Content: View>: View {
    // MARK: - Properties

    /// Legend text for the fieldset.
    public let legend: String

    /// Help text shown below the legend.
    public var hint: String?

    /// Error message for the group.
    public var error: String?

    /// Whether the fieldset is required.
    public var isRequired: Bool

    /// Whether the fieldset and all contents are disabled.
    public var isDisabled: Bool

    /// Content rendered inside the fieldset.
    public let content: Content

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        legend: String,
        hint: String? = nil,
        error: String? = nil,
        isRequired: Bool = false,
        isDisabled: Bool = false,
        @ViewBuilder content: () -> Content
    ) {
        self.legend = legend
        self.hint = hint
        self.error = error
        self.isRequired = isRequired
        self.isDisabled = isDisabled
        self.content = content()
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._1) {
            // 1. Legend
            legendView

            // 2. Hint
            if let hint, !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .accessibilityIdentifier("civ-hint")
            }

            // 3. Error
            if let error, !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityIdentifier("civ-error")
                    .accessibilityAddTraits(.updatesFrequently)
            }

            // 4. Content
            VStack(alignment: .leading, spacing: 0) {
                content
            }
        }
        .padding(.bottom, CivTokens.Spacing._4)
        .disabled(isDisabled)
        .opacity(isDisabled ? 0.5 : 1.0)
        .accessibilityElement(children: .contain)
    }

    // MARK: - Subviews

    private var legendView: some View {
        HStack(spacing: CivTokens.Spacing._0_5) {
            Text(legend)
                .font(.system(size: CivTokens.Typography.FontSize.lg,
                              weight: CivTokens.Typography.FontWeight.bold))
                .foregroundColor(adaptiveColor(
                    light: CivTokens.Colors.Base.darkest,
                    dark: CivTokens.DarkColors.Base.darkest
                ))

            if isRequired {
                Text("*")
                    .font(.system(size: CivTokens.Typography.FontSize.lg,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityLabel("required")
            }
        }
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivFieldset_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var street = ""
        @State private var city = ""
        @State private var state = ""
        @State private var zip = ""

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivFieldset(legend: "Mailing address", hint: "Enter your current mailing address") {
                        CivTextInput(label: "Street address", value: $street, isRequired: true)
                        CivTextInput(label: "City", value: $city, isRequired: true)
                        CivTextInput(label: "State", value: $state, isRequired: true)
                        CivTextInput(label: "ZIP code", value: $zip, isRequired: true, mask: .zip)
                    }

                    CivFieldset(legend: "Disabled section", isDisabled: true) {
                        CivTextInput(label: "Cannot edit", value: .constant(""))
                    }
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivFieldset.kt
// CivUI — CivFieldset for Jetpack Compose
// Accessible fieldset container with legend, hint, error, and disabled cascade.
// Renders: legend -> hint -> error -> content (Section 508 compliant)

package gov.civui.components

import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.i18n.CivLocale
import gov.civui.tokens.CivTokens

/**
 * CompositionLocal for cascading disabled state from a CivFieldset
 * to its children, mimicking native fieldset disabled behavior.
 */
val LocalCivFieldsetDisabled = compositionLocalOf { false }

/**
 * Accessible fieldset container for government applications.
 *
 * Provides a card-like container with legend, hint, error messaging,
 * and disabled cascade. Similar to a native HTML fieldset but
 * implemented as a Compose Column with border.
 *
 * Usage:
 * ```kotlin
 * CivFieldset(
 *     legend = "Mailing address",
 *     hint = "Enter your current mailing address",
 * ) {
 *     CivTextInput(label = "Street", value = street, onValueChange = { street = it })
 *     CivTextInput(label = "City", value = city, onValueChange = { city = it })
 * }
 * ```
 */
@Composable
fun CivFieldset(
    legend: String,
    modifier: Modifier = Modifier,
    hint: String? = null,
    error: String? = null,
    required: Boolean = false,
    disabled: Boolean = false,
    content: @Composable () -> Unit,
) {
    val isDark = isSystemInDarkTheme()

    val labelColor = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_
    val borderColor = if (isDark) CivTokens.DarkColors.Base.lighter else CivTokens.Colors.Base.lighter

    // Inherit disabled from parent fieldset if present
    val parentDisabled = LocalCivFieldsetDisabled.current
    val effectiveDisabled = disabled || parentDisabled

    Column(
        modifier = modifier
            .fillMaxWidth()
            .border(
                width = CivTokens.Border.Width.default_,
                color = borderColor,
                shape = RoundedCornerShape(CivTokens.Border.Radius.lg),
            )
            .padding(CivTokens.Spacing._4)
            .alpha(if (effectiveDisabled) 0.5f else 1f)
            .semantics {
                contentDescription = buildString {
                    append(legend)
                    if (required) append(", required")
                    if (effectiveDisabled) append(", disabled")
                    if (error != null) append(". Error: $error")
                }
                if (error != null) {
                    error(error)
                }
            },
    ) {
        // 1. Legend (rendered as bold heading-style text)
        CivLabel(
            label = legend,
            required = required,
            labelColor = labelColor,
            errorColor = errorColor,
        )

        // 2. Hint
        CivHint(text = hint, color = hintColor)

        // 3. Error
        CivError(text = error, color = errorColor)

        // 4. Content with disabled cascade
        CompositionLocalProvider(LocalCivFieldsetDisabled provides effectiveDisabled) {
            content()
        }
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivFieldset")
@Composable
private fun CivFieldsetPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var street by remember { mutableStateOf("") }
        var city by remember { mutableStateOf("") }

        CivFieldset(
            legend = "Mailing address",
            hint = "Enter your current mailing address",
        ) {
            CivTextInput(
                label = "Street address",
                value = street,
                onValueChange = { street = it },
                required = true,
            )
            CivTextInput(
                label = "City",
                value = city,
                onValueChange = { city = it },
                required = true,
            )
        }

        CivFieldset(
            legend = "Disabled section",
            disabled = true,
            error = "This section has an error",
        ) {
            CivTextInput(
                label = "Cannot edit",
                value = "",
                onValueChange = {},
            )
        }
    }
}
